cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -428,6 +429,264 @@ const FunctionDoc make_struct_doc{"Wrap Arrays into a StructArray",
+        out->value = MakeNullScalar(list(items->type()));
+      } else {
+        ARROW_ASSIGN_OR_RAISE(auto result, builder->Finish());
+        ARROW_ASSIGN_OR_RAISE(out->value, MakeScalar(list(items->type()), result));
+      }
+    } else { /* occurrence == FIRST || LAST */
+      bool from_back = (occurrence == MapLookupOptions::LAST);
+
+      ARROW_ASSIGN_OR_RAISE(int64_t key_match_idx,
+                            GetOneMatchingIndex(*keys, *query_key, &from_back));
+      if (key_match_idx != -1) {
+        ARROW_ASSIGN_OR_RAISE(out->value, items->GetScalar(key_match_idx));
+      } else {
+        out->value = MakeNullScalar(items->type());
+      }
+    }
+    return Status::OK();
+  }
+};
+
+Result<ValueDescr> ResolveMapLookupType(KernelContext* ctx,
+                                        const std::vector<ValueDescr>& descrs) {
+  const auto& options = OptionsWrapper<MapLookupOptions>::Get(ctx);
+  std::shared_ptr<DataType> type = descrs.front().type;
+  std::shared_ptr<DataType> item_type = checked_cast<const MapType&>(*type).item_type();
+  std::shared_ptr<DataType> key_type = checked_cast<const MapType&>(*type).key_type();
+
+  if (!options.query_key) {
+    return Status::Invalid("map_lookup: query_key can't be empty.");
+  } else if (!options.query_key->is_valid) {
+    return Status::Invalid("map_lookup: query_key can't be null.");
+  } else if (!options.query_key->type || !options.query_key->type->Equals(key_type)) {
+    return Status::TypeError(
+        "map_lookup: query_key type and Map key_type don't match. Expected "
+        "type: ",
+        *key_type, ", but got type: ", *options.query_key->type);
+  }
+
+  if (options.occurrence == MapLookupOptions::Occurrence::ALL) {
+    return ValueDescr(list(item_type), descrs.front().shape);
+  } else { /* occurrence == FIRST || LAST */
+    return ValueDescr(item_type, descrs.front().shape);
+  }
+}
+
+struct ResolveMapLookup {
+  KernelContext* ctx;
+  const ExecBatch& batch;
+  Datum* out;
+
+  template <typename KeyType>
+  Status Execute() {
+    if (batch[0].kind() == Datum::SCALAR) {
+      return MapLookupFunctor<KeyType>::ExecMapScalar(ctx, batch, out);
+    }
+    return MapLookupFunctor<KeyType>::ExecMapArray(ctx, batch, out);
+  }
+
+  template <typename KeyType>
+  enable_if_physical_integer<KeyType, Status> Visit(const KeyType& type) {
+    return Execute<KeyType>();
+  }
+
+  template <typename KeyType>
+  enable_if_decimal<KeyType, Status> Visit(const KeyType& type) {
+    return Execute<KeyType>();
+  }
+
+  template <typename KeyType>
+  enable_if_base_binary<KeyType, Status> Visit(const KeyType& type) {
+    return Execute<KeyType>();
+  }
+
+  template <typename KeyType>
+  enable_if_boolean<KeyType, Status> Visit(const KeyType& type) {
+    return Execute<KeyType>();
+  }
+
+  template <typename KeyType>
+  enable_if_same<KeyType, FixedSizeBinaryType, Status> Visit(const KeyType& key) {
+    return Execute<KeyType>();
+  }
```

Comment:
Does this work if it's just a normal overload? `Status Visit(const FixedSizeBinaryType& key)`